http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/RedirectHandler.java

```diff
@@ -0,0 +1,782 @@
+            }
+            if (ctx == originalChannelContext) {
+                originalChannelContext.fireChannelRead(msg);
+                ctx.close();
+            } else {
+                markEndOfMessage(ctx, (HttpContent) msg);
+            }
+        }
+    }
+
+    /**
+     * Write content to backend using existing channel.
+     *
+     * @param ctx               Channel handler context
+     * @param httpCarbonRequest Carbon request
+     * @param httpRequest       Http request
+     */
+    private void writeContentToExistingChannel(ChannelHandlerContext ctx, HTTPCarbonMessage httpCarbonRequest,
+            HttpRequest httpRequest) {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Use existing channel" + ctx.channel().id() + " to send the redirect request.");
+        }
+        ctx.channel().attr(Constants.ORIGINAL_REQUEST).set(httpCarbonRequest);
+        ctx.write(httpRequest);
+        ctx.writeAndFlush(LastHttpContent.EMPTY_LAST_CONTENT);
+    }
+
+    /**
+     * Mark the end of response and reset channel attributes.
+     *
+     * @param ctx         Channel context
+     * @param httpContent Http content
+     */
+    private void markEndOfMessage(ChannelHandlerContext ctx, HttpContent httpContent) throws Exception {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Mark end of the message and reset channel attributes for channel : " + ctx.channel().id());
+        }
+        targetRespMsg.addHttpContent(httpContent);
+        targetRespMsg.setEndOfMsgAdded(true);
+        targetRespMsg = null;
+        currentRedirectCount = 0;
+        TargetChannel targetChannel = ctx.channel().attr(Constants.TARGET_CHANNEL_REFERENCE).get();
+        try {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Return target channel : " + targetChannel.getChannel().id() + " back to its pool from "
+                        + "RedirectHandler. Currently in channel : " + ctx.channel().id());
+            }
+            Util.resetChannelAttributes(ctx);
+            Util.resetChannelAttributes(originalChannelContext);
+            if (!isIdleHandlerOfTargetChannelRemoved) {
+                if (targetChannel.getChannel().isActive()) {
+                    targetChannel.getChannel().pipeline().remove(Constants.IDLE_STATE_HANDLER);
+                    isIdleHandlerOfTargetChannelRemoved = true;
+                }
+            }
+            ConnectionManager.getInstance().returnChannel(targetChannel);
+            if (ctx != originalChannelContext) {
+                ctx.close();
+            }
+        } catch (Exception exception) {
+            LOG.error(
+                    "Error occurred while returning target channel " + targetChannel.getChannel().id() + " from current"
+                            + " channel" + ctx.channel().id() + " " + "to its pool in " + "markEndOfMessage",
+                    exception);
+            exceptionCaught(ctx, exception.getCause());
```

Comment:
Are we logging the exception twice?

Reply:
This is fixed in https://github.com/wso2/carbon-transports/pull/415